src/common/exif.cc

```diff
@@ -3231,6 +3237,76 @@ static void dt_set_xmp_dt_history(Exiv2::XmpData &xmpData, const int imgid, int
   xmpData["Xmp.darktable.history_end"] = history_end;
 }
 
+// add timestamps to XmpData.
+static void set_xmp_timestamps(Exiv2::XmpData &xmpData, const int imgid)
+{
+  sqlite3_stmt *stmt;
+
+  DT_DEBUG_SQLITE3_PREPARE_V2(
+      dt_database_get(darktable.db),
+      "SELECT id, import_timestamp, change_timestamp, export_timestamp, print_timestamp"
+      " FROM main.images"
+      " WHERE id = ?1",
+      -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+
+  if(sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    xmpData["Xmp.darktable.import_timestamp"] = sqlite3_column_int(stmt, 1);
+    xmpData["Xmp.darktable.change_timestamp"] = sqlite3_column_int(stmt, 2);
+    xmpData["Xmp.darktable.export_timestamp"] = sqlite3_column_int(stmt, 3);
+    xmpData["Xmp.darktable.print_timestamp"] = sqlite3_column_int(stmt, 4);
+  }
+  else
+  {
+    xmpData["Xmp.darktable.import_timestamp"] = -1;
+    xmpData["Xmp.darktable.change_timestamp"] = -1;
+    xmpData["Xmp.darktable.export_timestamp"] = -1;
+    xmpData["Xmp.darktable.print_timestamp"] = -1;
+  }
+  sqlite3_finalize(stmt);
+}
+
+// read timestamps from XmpData
+gboolean read_xmp_timestamps(Exiv2::XmpData &xmpData, const int imgid)
+{
+  gboolean all_ok = TRUE;
+  sqlite3_stmt *stmt;
+  Exiv2::XmpData::iterator pos;
+
+  const char *timestamps[] = { "change_timestamp", "export_timestamp", "print_timestamp" };
+  // Do not read for import_ts. It must be updated at each import.
+
+  const int nb_timestamps = sizeof(timestamps) / sizeof(char *);
+  int i;
+  char xmpkey[1024];
+  char query[1024];
+  char values[1024];
+  char *ptr_values = values ;
+
+  for (i = 0 ; i < nb_timestamps ; i++)
+  {
+    snprintf(xmpkey, sizeof(xmpkey), "Xmp.darktable.%s", timestamps[i]);
+    if((pos = xmpData.findKey(Exiv2::XmpKey(xmpkey))) != xmpData.end())
+    {
+      snprintf(ptr_values, sizeof(values) - strlen(values), " %s = %ld", timestamps[i], pos->toLong());
```

Comment:
This is wrong to me : `sizeof(values)` is static and equal to 1024 but the ptr_values will point not only at the start.

I would use a tmp char array to but the result of snprintf() and then use `g_strlcat(values, tmp, sizeof(values);`

Reply:
You're right, it's much better this way. I don't know why I got into this weird thing without thinking about the strcat() functions. :-(